File: ffc/backends/ufc/finite_element.py
# -*- coding: utf-8 -*-
# Code generation format strings for UFC (Unified Form-assembly Code)
# This code is released into the public domain.
#
# The FEniCS Project (http://www.fenicsproject.org/) 2018.

ufc_finite_element_declaration = """
ufc_finite_element* create_{factory_name}();
"""

ufc_finite_element_factory = """
// Code for element {factory_name}

int value_dimension_{factory_name}(int64_t i)
{{
  {value_dimension}
}}

int reference_value_dimension_{factory_name}(int64_t i)
{{
  {reference_value_dimension}
}}

int evaluate_reference_basis_{factory_name}(double *reference_values,
                                            int64_t num_points,
                                            const double *X)
{{
  {evaluate_reference_basis}
}}

int evaluate_reference_basis_derivatives_{factory_name}(double *reference_values,
                                          int64_t order, int64_t num_points,
                                          const double *X)
{{
  {evaluate_reference_basis_derivatives}
}}

int transform_reference_basis_derivatives_{factory_name}(
    double *values, int64_t order, int64_t num_points,
    const double *reference_values, const double *X, const double *J,
    const double *detJ, const double *K, int cell_orientation)
{{
  {transform_reference_basis_derivatives}
}}

void map_dofs_{factory_name}(double *values, const double *vals,
                   const double *coordinate_dofs, int cell_orientation,
                   const ufc_coordinate_mapping *cm)
{{
  {map_dofs}
}}

void tabulate_reference_dof_coordinates_{factory_name}(double *reference_dof_coordinates)
{{
  {tabulate_reference_dof_coordinates}
}}

ufc_finite_element* create_sub_element_{factory_name}(int64_t i)
{{
  {create_sub_element}
}}

ufc_finite_element* create_{factory_name}()
{{
  ufc_finite_element* element = malloc(sizeof(*element));

  element->signature = {signature};
  element->cell_shape = {cell_shape};
  element->topological_dimension = {topological_dimension};
  element->geometric_dimension = {geometric_dimension};
  element->space_dimension = {space_dimension};
  element->value_rank = {value_rank};
  element->value_dimension = value_dimension_{factory_name};
  element->value_size = {value_size};
  element->reference_value_rank = {reference_value_rank};
  element->reference_value_dimension = reference_value_dimension_{factory_name};
  element->reference_value_size = {reference_value_size};
  element->degree = {degree};
  element->family = {family};
  element->evaluate_reference_basis = evaluate_reference_basis_{factory_name};
  element->evaluate_reference_basis_derivatives = evaluate_reference_basis_derivatives_{factory_name};
  element->transform_reference_basis_derivatives = transform_reference_basis_derivatives_{factory_name};
  element->map_dofs = map_dofs_{factory_name};
  element->tabulate_reference_dof_coordinates = tabulate_reference_dof_coordinates_{factory_name};
  element->num_sub_elements = {num_sub_elements};
  element->create_sub_element = create_sub_element_{factory_name};
  element->create = create_{factory_name};

  return element;
}};

// End of code for element {factory_name}
"""

File: libs/ffc-factory/finite_element.cpp
// Copyright (C) 2017 Garth N. Wells
//
// This file is part of FFC.
//
// FFC is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// FFC is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with FFC. If not, see <http://www.gnu.org/licenses/>.

#include <iostream>
#include <memory>
#include <stdexcept>
#include <vector>

#include <pybind11/pybind11.h>
#include <pybind11/numpy.h>

#include <ufc.h>

namespace py = pybind11;

namespace ufc_wrappers
{
  // Interface function to
  // ufc::finite_element::evaluate_reference_basis
  py::array_t<double> evaluate_reference_basis(ufc::finite_element &instance,
                                               py::array_t<double> x);

  py::array_t<double>
  evaluate_reference_basis_derivatives(const ufc::finite_element &instance,
                                       std::size_t order, py::array_t<double> x);

  py::array_t<double>
  transform_reference_basis_derivatives(ufc::finite_element &instance,
                                        std::size_t order,
                                        py::array_t<double> reference_values,
                                        py::array_t<double> X,
                                        py::array_t<double> J,
                                        py::array_t<double> detJ,
                                        py::array_t<double> K,
                                        int orientation);

  // Remove from UFC
  py::array_t<double>
  evaluate_basis(ufc::finite_element &instance, std::size_t i,
                 py::array_t<double> x,
                 py::array_t<double> coordinate_dofs,
                 int cell_orientation);

  // Remove from UFC
  py::array_t<double> evaluate_basis_all(ufc::finite_element &instance,
                                         py::array_t<double> x,
                                         py::array_t<double> coordinate_dofs,
                                         int cell_orientation);
  // Remove from UFC
  py::array_t<double> evaluate_basis_derivatives(ufc::finite_element &instance,
                                                 std::size_t i,
                                                 std::size_t n,
                                                 py::array_t<double> x,
                                                 py::array_t<double> coordinate_dofs,
                                                 int cell_orientation);

  // Remove from UFC
  py::array_t<double>
  evaluate_basis_derivatives_all(ufc::finite_element &instance,
                                 std::size_t n,
                                 py::array_t<double> x,
                                 py::array_t<double> coordinate_dofs,
                                 int cell_orientation);

  double evaluate_dof(ufc::finite_element &instance,
                      std::size_t i,
                      const ufc::function& f,
                      py::array_t<double> coordinate_dofs,
                      int cell_orientation,
                      const ufc::cell& cell);

  py::array_t<double> evaluate_dofs(ufc::finite_element &instance,
                                    const ufc::function& f,
                                    py::array_t<double> coordinate_dofs,
                                    int cell_orientation,
                                    const ufc::cell& cell);

  py::array_t<double> interpolate_vertex_values(ufc::finite_element &instance,
                                                py::array_t<double> dof_values,
                                                py::array_t<double> coordinate_dofs,
                                                int cell_orientation);

  py::array_t<double> tabulate_dof_coordinates(ufc::finite_element &instance,
                                               py::array_t<double> coordinate_dofs);

  py::array_t<double>
  tabulate_reference_dof_coordinates(ufc::finite_element &instance);

  // Add the Python submodule 'finite_element'
  void finite_element(py::module& m)
  {
    // Wrap ufc classes
    py::class_<ufc::finite_element, std::shared_ptr<ufc::finite_element>>
      element(m, "finite_element");

    // Wrap ufc::shape enum
    py::enum_<ufc::shape> shape(m, "shape");
    shape.value("interval", ufc::shape::interval);
    shape.value("triangle", ufc::shape::triangle);
    shape.value("tetrahedon", ufc::shape::tetrahedron);
    shape.value("quadrilateral", ufc::shape::quadrilateral);
    shape.value("hexahedron", ufc::shape::hexahedron);
    shape.value("vertex", ufc::shape::vertex);

    // Expose ufc::finite_element functions
    element.def("signature", &ufc::finite_element::signature);
    element.def("cell_shape", &ufc::finite_element::cell_shape);
    element.def("topological_dimension", &ufc::finite_element::topological_dimension);
    element.def("geometric_dimension", &ufc::finite_element::geometric_dimension);
    element.def("space_dimension", &ufc::finite_element::space_dimension);
    element.def("value_rank", &ufc::finite_element::value_rank);
    element.def("value_dimension", &ufc::finite_element::value_dimension);
    element.def("value_size", &ufc::finite_element::value_size);
    element.def("reference_value_rank", &ufc::finite_element::reference_value_rank);
    element.def("reference_value_dimension", &ufc::finite_element::reference_value_dimension);
    element.def("reference_value_size", &ufc::finite_element::reference_value_size);
    element.def("degree", &ufc::finite_element::degree);
    element.def("family", &ufc::finite_element::family);

    element.def("evaluate_reference_basis", &ufc_wrappers::evaluate_reference_basis,
                "evaluate reference basis");
    element.def("evaluate_reference_basis_derivatives",
                &ufc_wrappers::evaluate_reference_basis_derivatives,
              "evaluate reference basis");
    element.def("transform_reference_basis_derivatives",
                &ufc_wrappers::transform_reference_basis_derivatives,
                "transform reference basis derivatives");

    element.def("evaluate_basis", &ufc_wrappers::evaluate_basis, "evaluate basis");
    element.def("evaluate_basis_all", &ufc_wrappers::evaluate_basis_all,
                "evaluate basis all");

    element.def("evaluate_basis_derivatives", &ufc_wrappers::evaluate_basis_derivatives,
                "evaluate basis derivatives");
    element.def("evaluate_basis_derivatives_all", &ufc_wrappers::evaluate_basis_derivatives_all,
                "evaluate basis derivatives all");

    element.def("evaluate_dof", &ufc_wrappers::evaluate_dof, "evaluate dof");
    element.def("evaluate_dofs", &ufc_wrappers::evaluate_dofs, "evaluate dofs");

    element.def("interpolate_vertex_values", &ufc_wrappers::interpolate_vertex_values,
                "interpolate vertex values");
    element.def("tabulate_dof_coordinates", &ufc_wrappers::tabulate_dof_coordinates,
                "tabulate dof coordinates");

    element.def("tabulate_reference_dof_coordinates",
                &ufc_wrappers::tabulate_reference_dof_coordinates,
                "tabulate reference dof coordinates");

    element.def("num_sub_elements", &ufc::finite_element::num_sub_elements);

    element.def("create_sub_element", &ufc::finite_element::create_sub_element,
                py::return_value_policy::take_ownership);
    element.def("create", &ufc::finite_element::create,
                py::return_value_policy::take_ownership);
  }

}
